File: ai_hiring_platform/frontend/src/api/companyChat.ts
import { api, API_V1 } from './client';
import type { CompanyChatResponse, CompanyStoreStatus } from '../types';

/**
 * Client for the company knowledge base (`/api/v1/company-chat`).
 *
 * Kept separate from `api/chat.ts` because it talks to a different corpus with a
 * different response shape. The streaming helper mirrors `askChatStreaming` exactly —
 * same SSE framing, same fall back to the plain request — so both chat modes behave
 * identically to the user even though nothing is shared between the two pipelines.
 */

/** Ask one question against the company database. */
export async function askCompany(
  message: string,
  session_id: string,
  limit = 5,
  use_llm = true,
): Promise<CompanyChatResponse> {
  const res = await api.post('/company-chat/query', { message, session_id, limit, use_llm });
  if (res.data && res.data.success) {
    return res.data.data as CompanyChatResponse;
  }
  throw new Error(res.data?.message || 'Company request failed');
}

/**
 * Ask a question and receive real pipeline progress while it runs.
 *
 * Server-Sent Events over POST, so `fetch` + a stream reader rather than EventSource
 * (which is GET-only). Falls back to the plain request when streaming is unavailable.
 */
export async function askCompanyStreaming(
  message: string,
  session_id: string,
  onStage: (stage: string, detail: string) => void,
  limit = 5,
  use_llm = true,
): Promise<CompanyChatResponse> {
  let response: Response;
  try {
    response = await fetch(`${API_V1}/company-chat/stream`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ message, session_id, limit, use_llm }),
    });
  } catch {
    return askCompany(message, session_id, limit, use_llm);
  }
  if (!response.ok || !response.body) {
    return askCompany(message, session_id, limit, use_llm);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';
  let result: CompanyChatResponse | null = null;

  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });

    // SSE frames are separated by a blank line; keep any partial frame buffered.
    const frames = buffer.split('\n\n');
    buffer = frames.pop() ?? '';

    for (const frame of frames) {
      const eventLine = frame.split('\n').find((l) => l.startsWith('event:'));
      const dataLine = frame.split('\n').find((l) => l.startsWith('data:'));
      if (!dataLine) continue;
      const payload = JSON.parse(dataLine.slice(5).trim());
      const kind = eventLine?.slice(6).trim();
      if (kind === 'stage') onStage(payload.stage, payload.detail);
      else if (kind === 'result') result = payload as CompanyChatResponse;
      else if (kind === 'error') throw new Error(payload.message || 'The assistant failed.');
    }
  }

  if (!result) throw new Error('The assistant closed the connection without answering.');
  return result;
}

/** Clear the company conversation's follow-up memory on the backend. */
export async function resetCompanyChat(session_id: string): Promise<void> {
  await api.post('/company-chat/reset', null, { params: { session_id } });
}

/**
 * Whether the company database is configured, reachable and populated.
 *
 * Read before offering the toggle, so a recruiter is told the store is empty rather
 * than being handed an assistant that answers nothing.
 */
export async function getCompanyStatus(): Promise<CompanyStoreStatus> {
  const res = await api.get('/company-chat/status');
  if (res.data && res.data.success) {
    return res.data.data as CompanyStoreStatus;
  }
  throw new Error(res.data?.message || 'Could not read company store status');
}

File: ai_hiring_platform/backend/app/api/v1/routers/company_chat.py
"""
Company knowledge base endpoints (`/api/v1/company-chat`).

Thin by design (Golden Rule 7): every endpoint delegates to `services/company/*` and
wraps the result in the standard `{success, message, data}` envelope. Nothing here
touches the resume pipeline — this router is reached only when the chat UI is in
company mode.
"""
import json
import queue
import threading
from typing import Optional

from fastapi import APIRouter, Query
from fastapi.responses import StreamingResponse

from app.core.logging import logger
from app.schemas.company_chat import (
    CompanyChatRequest,
    CompanyChatResponse,
    CompanyStoreStatus,
)
from app.schemas.response import ApiResponse
from app.services.company import company_chat, company_retrieval, qdrant_store

router = APIRouter()


@router.post("/query", response_model=ApiResponse[CompanyChatResponse])
def company_query(payload: CompanyChatRequest):
    """Answer one question against the company knowledge base."""
    logger.info(f"Company query [session={payload.session_id}]: {payload.message[:160]!r}")
    result = company_chat.answer(
        message=payload.message,
        session_id=payload.session_id,
        limit=payload.limit,
        use_llm=payload.use_llm,
    )
    return ApiResponse[CompanyChatResponse](
        success=True,
        message="Refused." if result["refused"] else "Answer generated.",
        data=CompanyChatResponse(**result),
    )


@router.post("/stream")
def company_query_stream(payload: CompanyChatRequest):
    """
    The same answer, streaming real pipeline stages as Server-Sent Events.

    Mirrors `/chat/stream` so the UI can drive both modes with one component: the work
    runs on a worker thread and pushes a `stage` event as each step actually begins,
    then a final `result` event carrying the same payload as `/query`.
    """
    logger.info(f"Company stream [session={payload.session_id}]: {payload.message[:160]!r}")
    events: "queue.Queue[Optional[str]]" = queue.Queue()

    def on_progress(stage: str, detail: str) -> None:
        events.put("event: stage\ndata: " + json.dumps({"stage": stage, "detail": detail}) + "\n\n")

    def work() -> None:
        try:
            result = company_chat.answer(
                message=payload.message,
                session_id=payload.session_id,
                limit=payload.limit,
                use_llm=payload.use_llm,
                progress=on_progress,
            )
            body = CompanyChatResponse(**result).model_dump()
            events.put("event: result\ndata: " + json.dumps(body, default=str) + "\n\n")
        except Exception as e:
            logger.error(f"Company stream failed: {e}", exc_info=True)
            events.put("event: error\ndata: " + json.dumps({"message": str(e)}) + "\n\n")
        finally:
            events.put(None)

    threading.Thread(target=work, daemon=True).start()

    def generate():
        while True:
            item = events.get()
            if item is None:
                break
            yield item

    return StreamingResponse(
        generate(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


@router.post("/reset", response_model=ApiResponse[dict])
def company_reset(session_id: str = Query("default")):
    """Forget a company conversation's follow-up memory."""
    company_chat.reset_session(session_id)
    return ApiResponse[dict](
        success=True, message="Conversation reset.", data={"session_id": session_id}
    )


@router.get("/status", response_model=ApiResponse[CompanyStoreStatus])
def company_status():
    """
    Whether the company database is configured, reachable and populated.

    The UI reads this before offering the company toggle, so a recruiter is told the
    store is empty rather than being handed an assistant that answers nothing.
    """
    return ApiResponse[CompanyStoreStatus](
        success=True,
        message="Company store status retrieved.",
        data=CompanyStoreStatus(**company_chat.store_status()),
    )


@router.get("/companies", response_model=ApiResponse[list])
def company_list():
    """Every company in the store, for a picker or a sanity check after loading."""
    if not qdrant_store.configured():
        return ApiResponse[list](success=True, message="Not configured.", data=[])
    return ApiResponse[list](
        success=True,
        message="Companies retrieved.",
        data=qdrant_store.list_companies(),
    )


@router.post("/refresh", response_model=ApiResponse[dict])
def company_refresh():
    """
    Re-read the name index after a load, without restarting the API.

    The loader runs as a separate script, so a running server would otherwise keep a
    stale list of company and people names for up to the cache TTL.
    """
    company_retrieval.refresh_index()
    return ApiResponse[dict](
        success=True, message="Company index refreshed.", data=qdrant_store.status()
    )
